### PR TITLE
`Number_mapped_unique_molecules` is now calculated on the genome BAM

### PR DESCRIPTION
**Pull request by ctuni**
On the QC table, we have two metrics related to the number of unique mapped molecules (the number and the percentage). Those statistics were calculated on the transcriptome BAM, but should be calculated on the genome BAM. This fix should solve the issue.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206777252434507

### CODE REVIEW

**Comment by julienlag** [comment]
please post test results here @ctuni , including EXP-116 if possible, thanks

**Comment by ctuni** [comment]
Output directory for a test with the full FL116 samplesheet is here: `/mnt/efs/home/ctuni/test_umi/out/2024_03_11_15_04_43/b37cced4`

Those are the values with the new branch (genome_bam) compared to the values from the dev branch (transcriptome_bam, `FL-ANALYSIS-4aeef832`) :

	
</head>

<body>

Sample | Number_mapped_unique_molecules_GENOME_BAM | Number_mapped_unique_molecules_TRANSCRIPTOME_BAM
-- | -- | --
SAMP231_EXP116_A1 | 244085 | 157495
SAMP232_EXP116_A2 | 235179 | 189752
SAMP233_EXP116_A3 | 187274 | 129568
SAMP234_EXP116_A4 | 226429 | 136906
SAMP235_EXP116_C1 | 277885 | 192594
SAMP236_EXP116_C2 | 315053 | 223072
SAMP237_EXP116_C3 | 252813 | 62788
SAMP238_EXP116_C4 | 254869 | 79351


</body>

</html>


They are substantially different, especially for samples 237 and 238 (samples from FL082 plasma processed by Sol, they have a lower exonic count compared to the others).

I think it makes sense to calculate this metric on the genome BAM instead of the transcriptome BAM because we'll be using that number of "unique molecules" downstream to asses whether the exonic, intronic, biotype count, etc make sense.

File: workflows/rnaseq.nf [hunk 1]
```diff
@@ -834,7 +834,7 @@ workflow RNASEQ {
     ch_Flomics_UMI_dedup_rate_QC = Channel.empty()
 
     if (params.with_umi) {
-        ch_bams_umi_dedup= DEDUP_UMI_UMITOOLS_TRANSCRIPTOME.out.bam.join(ALIGN_STAR.out.bam_transcript)
+        ch_bams_umi_dedup = DEDUP_UMI_UMITOOLS_GENOME.out.bam.join(ALIGN_STAR.out.bam)
```
Comment (julienlag) [inline workflows/rnaseq.nf:846]: @ctuni are you sure this affects *only* the deduplication stats calculation? And not, for example, the deduplication before salmon quantification, which should be performed on the transcriptome bam, if I remember correctly? 
Can this be visualized in the DAG?

**Comment by julienlag** [comment]
the differences in raw counts that you're observing are expected. What about percentages? Could you please specify what is the denominator in this calculation, in both cases (with pointers to code), thanks!

> Output directory for a test with the full FL116 samplesheet is here: `/mnt/efs/home/ctuni/test_umi/out/2024_03_11_15_04_43/b37cced4`
> 
> Those are the values with the new branch (genome_bam) compared to the values from the dev branch (transcriptome_bam, `FL-ANALYSIS-4aeef832`) :
> Sample 	Number_mapped_unique_molecules_GENOME_BAM 	Number_mapped_unique_molecules_TRANSCRIPTOME_BAM
> SAMP231_EXP116_A1 	244085 	157495
> SAMP232_EXP116_A2 	235179 	189752
> SAMP233_EXP116_A3 	187274 	129568
> SAMP234_EXP116_A4 	226429 	136906
> SAMP235_EXP116_C1 	277885 	192594
> SAMP236_EXP116_C2 	315053 	223072
> SAMP237_EXP116_C3 	252813 	62788
> SAMP238_EXP116_C4 	254869 	79351
> 
> They are substantially different, especially for samples 237 and 238 (samples from FL082 plasma processed by Sol, they have a lower exonic count compared to the others).
> 
> I think it makes sense to calculate this metric on the genome BAM instead of the transcriptome BAM because we'll be using that number of "unique molecules" downstream to asses whether the exonic, intronic, biotype count, etc make sense.

**Comment by ctuni** [comment]
Yes! The denominator is the number of lines on the "normal" out of STAR BAM file. In the `dev` version the number of lines in the UMI-deduped transcriptome BAM is divided by the number of lines in the normal transcriptome BAM and multiplied by 100. In this branch, both BAM files are the genome BAM (deduped and "normal"). This is done in [this script](https://github.com/Flomics/rnaseq/blob/dev/bin/umitools_dedup_rate.sh).
So the changes in percentages are not as striking:
<!DOCTYPE html>

<html>
<head>
	
	
	
</head>

<body>

Sample | Percentage_unique_molecules_GENOME_BAM | Percentage_unique_molecules_TRANSCRIPTOME_BAM
-- | -- | --
SAMP231_EXP116_A1 | 90.8 | 89.9
SAMP232_EXP116_A2 | 90.8 | 90.1
SAMP233_EXP116_A3 | 77.3 | 75.6
SAMP234_EXP116_A4 | 82 | 79.1
SAMP235_EXP116_C1 | 95.1 | 94.3
SAMP236_EXP116_C2 | 94 | 93.2
SAMP237_EXP116_C3 | 97.5 | 96.3
SAMP238_EXP116_C4 | 95.5 | 92.6


</body>

</html>

**Comment by julienlag** [comment]
> Yes! The denominator is the number of lines on the "normal" out of STAR BAM file. In the `dev` version the number of lines in the UMI-deduped transcriptome BAM is divided by the number of lines in the normal transcriptome BAM and multiplied by 100. In this branch, both BAM files are the genome BAM (deduped and "normal"). This is done in [this script](https://github.com/Flomics/rnaseq/blob/dev/bin/umitools_dedup_rate.sh). So the changes in percentages are not as striking:
> Sample 	Percentage_unique_molecules_GENOME_BAM 	Percentage_unique_molecules_TRANSCRIPTOME_BAM
> SAMP231_EXP116_A1 	90.8 	89.9
> SAMP232_EXP116_A2 	90.8 	90.1
> SAMP233_EXP116_A3 	77.3 	75.6
> SAMP234_EXP116_A4 	82 	79.1
> SAMP235_EXP116_C1 	95.1 	94.3
> SAMP236_EXP116_C2 	94 	93.2
> SAMP237_EXP116_C3 	97.5 	96.3
> SAMP238_EXP116_C4 	95.5 	92.6

ok these stats look good to me

**Comment by ctuni** [comment]
Perfect! Can we merge it? :smile:

File: workflows/rnaseq.nf (hunk 1 above)
Comment (julienlag) [reply workflows/rnaseq.nf:846]: please address this before merging @ctuni :-)

**Review by julienlag — CHANGES_REQUESTED** [review]

**Comment by ctuni** [comment]
I don't have the DAG yet but the change should not affect any other UMI-related number. The change introduced by this PR is on line 846 after everything on the pipeline has been done except the Flomics QC module. The modification only changes the input files (from umi deduped transcriptome bam to umi deduped genome bam) for the `FLOMICS_UMI_DEDUP_QC` module. No input file for SALMON or other important modules are changed.

**Comment by ctuni** [comment]
From the DOT:
```
v577 [label="NFCORE_RNASEQ:RNASEQ:FLOMICS_UMI_DEDUP_QC:FLOMICS_QC_CALCULATE_UMI_DEDUP_RATE"];
v576 -> v577 [label="bam_files"];

v577 [label="NFCORE_RNASEQ:RNASEQ:FLOMICS_UMI_DEDUP_QC:FLOMICS_QC_CALCULATE_UMI_DEDUP_RATE"];
v578 [shape=circle,label="",fixedsize=true,width=0.1,xlabel="collect"];
v577 -> v578 [label="umi_dedup_rate"];

v578 [shape=circle,label="",fixedsize=true,width=0.1,xlabel="collect"];
v579 [shape=point];
v578 -> v579 [label="ch_Flomics_UMI_dedup_rate_QC"];

v577 [label="NFCORE_RNASEQ:RNASEQ:FLOMICS_UMI_DEDUP_QC:FLOMICS_QC_CALCULATE_UMI_DEDUP_RATE"];
v580 [shape=circle,label="",fixedsize=true,width=0.1,xlabel="collect"];
v577 -> v580 [label="umi_dedup_rate"];

v580 [shape=circle,label="",fixedsize=true,width=0.1,xlabel="collect"];
v639 [label="NFCORE_RNASEQ:RNASEQ:FLOMICS_QC:FLOMICS_QC_AGGREGATOR"];
v580 -> v639 [label="umi_dedup_rate_data"];
```

The process `FLOMICS_UMI_DEDUP_QC` appears for the first time in `v577`. And then its output `umi_dedup_rate` (`v580`) only points to the aggregator process `v639`.

The change in the file used to calculate the "umi dedup rate" does not affect upstream deduplications used by Salmon and others

**Pull request merged**